src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -8,14 +8,230 @@
+    auto res = compiler->getDxcUtils()->CreateBlob(reinterpret_cast<const void*>(source.data()), source.size(), CP_UTF8, &src);
+    assert(SUCCEEDED(res));
+
+    DxcBuffer sourceBuffer;
+    sourceBuffer.Ptr = src->GetBufferPointer();
+    sourceBuffer.Size = src->GetBufferSize();
+    sourceBuffer.Encoding = 0;
+
+    ComPtr<IDxcResult> compileResult;
+    res = compiler->getDxcCompiler()->Compile(&sourceBuffer, args, argCount, nullptr, IID_PPV_ARGS(compileResult.GetAddressOf()));
+    // If the compilation failed, this should still be a successful result
+    assert(SUCCEEDED(res));
+
+    HRESULT compilationStatus = 0;
+    res = compileResult->GetStatus(&compilationStatus);
+    assert(SUCCEEDED(res));
+
+    ComPtr<IDxcBlobEncoding> errorBuffer;
+    res = compileResult->GetErrorBuffer(errorBuffer.GetAddressOf());
+    assert(SUCCEEDED(res));
+
+    DxcCompilationResult result;
+    result.errorMessages = errorBuffer;
+    result.compileResult = compileResult;
+    result.objectBlob = nullptr;
+
+    if (!SUCCEEDED(compilationStatus))
+    {
+        options.preprocessorOptions.logger.log(result.GetErrorMessagesString(), system::ILogger::ELL_ERROR);
+        return result;
+    }
+
+    ComPtr<IDxcBlob> resultingBlob;
+    res = compileResult->GetResult(resultingBlob.GetAddressOf());
+    assert(SUCCEEDED(res));
+
+    result.objectBlob = resultingBlob;
+
+    return result;
+}
+
+
+std::string CHLSLCompiler::preprocessShader(std::string&& code, IShader::E_SHADER_STAGE& stage, const SPreprocessorOptions& preprocessOptions) const
+{
+    if (preprocessOptions.extraDefines.size())
+    {
+        insertExtraDefines(code, preprocessOptions.extraDefines);
+    }
+    if (preprocessOptions.includeFinder != nullptr)
+    {
+        IShaderCompiler::disableAllDirectivesExceptIncludes(code);
+
+        tcpp::StringInputStream codeIs = tcpp::StringInputStream(code);
+        tcpp::Lexer lexer(codeIs);
+        tcpp::Preprocessor proc(
+            lexer,
+            [&](auto errorInfo) {
+                preprocessOptions.logger.log("Pre-processor error at line %i:\n%s", nbl::system::ILogger::ELL_ERROR, errorInfo.mLine, tcpp::ErrorTypeToString(errorInfo.mType).c_str());
+            },
+            [&](auto path, auto isSystemPath) {
+                return getInputStreamInclude(
+                    preprocessOptions.includeFinder, m_system.get(), preprocessOptions.maxSelfInclusionCount + 1u,
+                    path.c_str(), !isSystemPath
+                );
+            }
+        );
+
+        auto pragmaShaderStageCallback = [&](IShader::E_SHADER_STAGE _stage) {
+            return [&](tcpp::Preprocessor& preprocessor, tcpp::Lexer& lexer, const std::string& text) {
+                stage = _stage;
+                return std::string("");
+            };
+        };
+
+        proc.AddCustomDirectiveHandler(std::string("pragma shader_stage"), [&](tcpp::Preprocessor& preprocessor, tcpp::Lexer& lexer, const std::string& text) {
```

Comment:
you should still do this when preprocessOptions.includeFinder is nullptr. 
https://github.com/Devsh-Graphics-Programming/Nabla/pull/437/files#r1089687024